### PR TITLE
Prevent jumping of the player and wrong padding on devices with cutout

### LINKED ISSUE #4040: Content not centered with 18:9 screen -> weird behaviour with navigation bar
### Version

- d8b5549fd98e43f18b993c5f36f2385cf9a80620 

### Steps to reproduce the bug
1. Look for a vertical video, eg [this one](https://www.youtube.com/watch?v=KRntP-q_R9s)
2. Launch fullscreen

### Expected behavior

It is centered, and there the video doesn't "jump" to give space for navigation bar
### Actual behaviour

It's not centered, then when you click, the video goes up so that there is space for navigation bar
### Screenshots/Screen recordings

0.19.8 
0 19 8_nav_bar_1 0 19 8_nav_bar_2

unified
unified_nav_bar_1 unified_nav_bar_2

gif for this "jump" with unified

### PR DESCRIPTION
**Pull request by avently**
#### What is it?
- [x] Bug fix (user facing)

#### Description of the changes in your PR
On devices with cutout there is a "jump" when you are in multiWindow mode or when you're watching vertical video in fullscreen. This PR fixes that jump (the first commit).

On the same devices there is a problem related to top padding on vertical videos in portrait in fullscreen. This is more interesting because a couple hours ago I thought I will not be able to fix it. This problem makes situation when the center of the player is not in the center of the screen. After hours trying I installed the app on my phone and the issue is fixed actually (same Android version but different result). So I have no idea will my method (from the second commit) work on user's devices or not. I think the issue is still here but on devices with non-standard screen ratio. But I don't know how to fix it. You can merge the PR as is or merge only first commit, or drop it.

In short:
- in multiwindow mode all videos work great
- there is no jump that happened before when a user shows/hides controls in the player
- there is a correct padding on top of the player (but I'm sure it isn't on devices with non-standart ratio).

Everyone who has a device with a hole, cutout, whatever check this build in vertical videos in landscape, in portrait, in multi window mode, check non-vertical videos too.
 
#### Fixes the following issue(s)
partially fixes #4040

#### Testing apk
[app-release-final.zip](https://github.com/TeamNewPipe/NewPipe/files/5085735/app-release-final.zip)

#### Agreement
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

### CODE REVIEW

File: app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java [hunk 1]
```diff
@@ -1565,6 +1571,20 @@ private int getStatusBarHeight() {
...
+    private boolean hasCutout(final float statusBarHeight) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+            final DisplayMetrics metrics = getRootView().getResources().getDisplayMetrics();
+            final float defaultStatusBarHeight = TypedValue.applyDimension(
+                    TypedValue.COMPLEX_UNIT_DIP, 25, metrics);
+            return statusBarHeight > defaultStatusBarHeight;
+        }
+        return false;
+    }
+
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java:1587]: That looks like it should be in DeviceUtils.

**Comment by blackbox87** [comment]
> This problem makes situation when the center of the player is not in the center of the screen. After hours trying I installed the app on my phone and the issue is fixed actually (same Android version but different result). So I have no idea will my method (from the second commit) work on user's devices or not. I think the issue is still here but on devices with non-standard screen ratio. But I don't know how to fix it. You can merge the PR as is or merge only first commit, or drop it.

That commit actually makes the player off-center for me while using a Poco F1. It was perfect before the commit as it matched the official YouTube app, minus pinch to zoom.

YouTube:
YouTube

NewPipe v0.19.8:
NewPipe

NewPipe Cutout:
NewPipe_Cutout

**Comment by blackbox87** [comment]
> This problem makes situation when the center of the player is not in the center of the screen. After hours trying I installed the app on my phone and the issue is fixed actually (same Android version but different result). So I have no idea will my method (from the second commit) work on user's devices or not. I think the issue is still here but on devices with non-standard screen ratio. But I don't know how to fix it. You can merge the PR as is or merge only first commit, or drop it.

That commit actually makes the player off-center for me while using a Poco F1. It was perfect before the commit as it was almost identical to the official YouTube app, minus pinch to zoom.

Aligning to the notch means you'll probably lose some of the image. Like in my case it's now off-center and I lose some of the image on the right side of my screen due to the phones rounded corners.

YouTube:
YouTube

NewPipe v0.19.8:
NewPipe

NewPipe Cutout:
NewPipe_Cutout

**Comment by avently** [comment]
@blackbox87 
Thank you for testing. Actually we need to compare the apk from this PR with the apk from this post: https://github.com/TeamNewPipe/NewPipe/projects/17#card-42918515

You compared with 0.19.8 which doesn't make sense. In 0.19.8 the app uses default mechanism of handling the cutout, everything is done by Android. On the newest version there is no way to give this job to Android because of the need to use Fragment instead of an Activity. 

So, just compare with the latest apk from the link. More info about the problem the PR trying to fix can be found in #4040

**Comment by blackbox87** [comment]
> You compared with 0.19.8 which doesn't make sense. In 0.19.8 the app uses default mechanism of handling the cutout, everything is done by Android. On the newest version there is no way to give this job to Android because of the need to use Fragment instead of an Activity.

Sorry about that. I haven't been testing debug versions for a while now, so I didn't think to compare it to a different debug build. But I can confirm that the builds by @B0pol all have the same issue. It's awful.

When a device is in landscape you're not supposed to align to the cutcout because they come in many different sizes. If you do then you'll likely end up with off-centered videos and only having rounded corners on one side of your screen. That's why YouTube, VLC & many other players don't do anything about the notch and always center the videos to fit.

**Comment by blackbox87** [comment]
> You compared with 0.19.8 which doesn't make sense. In 0.19.8 the app uses default mechanism of handling the cutout, everything is done by Android. On the newest version there is no way to give this job to Android because of the need to use Fragment instead of an Activity.

Sorry about that. I haven't been testing debug versions for a while now, so I didn't think to compare it to a different debug build. But I can confirm that the builds by @B0pol all have the same issue. It's awful.

When a device is in landscape you're not supposed to align videos to the cutcout because cutcouts come in many different sizes. If you do then you'll likely end up with off-centered videos and only having rounded corners on one side of your screen. That's why YouTube, VLC & many other players don't do anything about the notch and always center the videos to fit.

**Comment by blackbox87** [comment]
> You compared with 0.19.8 which doesn't make sense. In 0.19.8 the app uses default mechanism of handling the cutout, everything is done by Android. On the newest version there is no way to give this job to Android because of the need to use Fragment instead of an Activity.

Sorry about that. I haven't been testing debug versions for a while now, so I didn't think to compare it to a different debug build. But I can confirm that the builds by @B0pol all have the same issue. It's awful.

You're not supposed to align videos to the cutcout while using landscape because cutcouts come in many different sizes. If you do then you'll likely end up with off-centered videos and only having rounded corners on only one side of your screen. That's why YouTube, VLC & many other players don't do anything about the notch and always center the videos to fit.

**Comment by blackbox87** [comment]
> You compared with 0.19.8 which doesn't make sense. In 0.19.8 the app uses default mechanism of handling the cutout, everything is done by Android. On the newest version there is no way to give this job to Android because of the need to use Fragment instead of an Activity.

Sorry about that. I haven't been testing debug versions for a while now, so I didn't think to compare it to a different debug build. But I can confirm that the builds by @B0pol all have the same issue. It's awful.

You're not supposed to align videos to the cutcout while using landscape because cutcouts come in many different sizes. If you do then you'll likely end up with off-centered videos and only having rounded corners on one side of your screen. That's why YouTube, VLC & many other players don't do anything about the notch and always center the videos to fit.

File: app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java (hunk 1 above)
Comment (avently) [reply app/src/main/java/org/schabi/newpipe/player/VideoPlayerImpl.java]: Force-pushed (which was a bad idea but anyway) a commit with a new place for this code

File: app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java [hunk 2]
```diff
@@ -1995,6 +2000,11 @@ private void hideSystemUi() {
             return;
         }
 
+        // Prevent jumping of the player on devices with cutout
```
Comment (blackbox87) [inline app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:2003]: Change to the following and it fixes landscape too.
```java
        // Prevent jumping of the player on devices with cutout
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            if (isLandscape()) {
                activity.getWindow().getAttributes().layoutInDisplayCutoutMode =
                        WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES;
            } else {
                activity.getWindow().getAttributes().layoutInDisplayCutoutMode =
                        WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_NEVER;
            }
        }
```

File: app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java [hunk 3]
```diff
@@ -1995,6 +2000,11 @@ private void hideSystemUi() {
...
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+            activity.getWindow().getAttributes().layoutInDisplayCutoutMode =
+                    WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_NEVER;
+        }
```
Comment (blackbox87) [inline app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:2007]: Change to the following and it fixes landscape too.
```java
        // Prevent jumping of the player on devices with cutout
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            if (isLandscape()) {
                activity.getWindow().getAttributes().layoutInDisplayCutoutMode =
                        WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES;
            } else {
                activity.getWindow().getAttributes().layoutInDisplayCutoutMode =
                        WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_NEVER;
            }
        }
```

Comment (avently) [reply app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:2007]: I can do this but in this case a title and a description will be hidden under possible cutout from top right corner (in landscape orientation it will be in top left corner). I'm not sure about how much devices have such cutout but it's not an ideal choice anyway.

Comment (blackbox87) [reply app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:2007]: That's actually a separate issue that's caused by the status bar height and navbar height not being accounted for while viewing a video in landscape. So I'm currently trying to work out the best way to center the controls.

Comment (blackbox87) [reply app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:2007]: That's actually a separate issue that's caused by the status bar height (and navbar height?) not being accounted for while viewing a video in landscape. So I'm currently trying to work out the best way to center the controls.

Comment (avently) [reply app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:2007]: Good to know that someone cares about it. Not only landscape mode has the problem with the centering but vertical videos too.

**Comment by blackbox87** [comment]
@avently [Check out my commit](https://github.com/blackbox87/NewPipe/commits/video-placement). It should give you perfectly centered videos and all of the controls will have the correct padding.

And although it's unrelated to this problem I noticed that screen rotation doesn't always behave correctly.

1. Enable auto-rotate and watch a video in landscape
2. Disable auto-rotate while watching the video in landscape

Previously NewPipe would return you to your default orientation, but now you'll remain stuck in landscape and NewPipe's resize button won't do anything no matter how many times you press on it. So as it seems buggy I made [this change](https://github.com/blackbox87/NewPipe/commits/auto-rotate-fix) which restores the original behaviour.

**Comment by blackbox87** [comment]
@avently [Check out my commit](https://github.com/blackbox87/NewPipe/commits/video-placement). It should give you perfectly centered videos and all of the controls will have the correct padding.

And although it's unrelated to the bug reports mentioned here I noticed that screen rotation doesn't always behave correctly.

1. Enable auto-rotate and watch a video in landscape
2. Disable auto-rotate while watching the video in landscape

Previously NewPipe would return you to your default orientation, but now you'll remain stuck in landscape and NewPipe's resize button won't do anything no matter how many times you press on it. Maybe that's the intended behaviour now, but as it seems buggy to me I made [this change](https://github.com/blackbox87/NewPipe/commits/auto-rotate-fix) which restores the original behaviour.

**Comment by avently** [comment]
@blackbox87 I didn't look closely on your changes but looks like the changes depend on this PR. And you say now everything is fine. So we can merge this PR and then to wait your PR, right? 

About autorotation issue: can you make a PR too?

**Comment by blackbox87** [comment]
> So we can merge this PR and then to wait your PR, right?

It could be done that way or you could just copy my changes into your branch. I'm happy with whatever as long as this gets fixed.

> About autorotation issue: can you make a PR too?

I'd like to know if it's the intended behaviour or not first. Maybe @Stypox or @TobiGr can comment?

It's possible that they want NewPipe to remain in landscape mode, but having the button not do anything when you disable auto-rotate seems like a bug either way. It should always be able to toggle between portrait and landscape.

**Comment by avently** [comment]
@blackbox87 I prefer to maintain only my code. So better to have your code inside your PR.

About autorotation: this is my code that makes such situation so if your solution fixes it just make a PR. For me it was like a rare case (when you lock orientation while in landscape) and I just skipped that case

**Comment by blackbox87** [comment]
> @blackbox87 I prefer to maintain only my code. So better to have your code inside your PR.

Fair enough. If someone merges this in then I'll update my fork and create a pull request.

> About autorotation: this is my code that makes such situation so if your solution fixes it just make a PR. For me it was like a rare case (when you lock orientation while in landscape) and I just skipped that case

Well there's two ways to fix the issue.

1. Set the orientation to portrait if auto-rotate is disabled and certain conditions are met.
2. Fix the resize button so that you can always toggle between portrait and landscape.

I'm not familiar with the code, which is why I went with the first option and made NewPipe behave as it used to. But I think fixing the button might actually be the better solution.

**Comment by avently** [comment]
@blackbox87 second way is not an option at all. So, yeah, the first way is what we need.

**Review by Stypox — APPROVED** [review]
Code looks good, though I don't have a device to test this on. But if @blackbox87 says he has a fully working solution based on this one I'd merge right away, also to prevent rebasing issues. Thank you @avently ;-) 

@blackbox87 thank you for your investigations, as soon as this PR will get merged feel free to create two separate PRs for cutout and rotation. I think your rotation approach is right, obviously having a button doing nothing is a bad idea and the previous behaviour was ok. Thanks :-D

**Comment by blackbox87** [comment]
@Stypox 
> I think your rotation approach is right, obviously having a button doing nothing is a bad idea and the previous behaviour was ok.

I've actually figured out how to fix the button, although @avently said that it wouldn't be an option. So we could have NewPipe remain in landscape and only change to portrait when the fullscreen button is pressed again. Personally I think that might be the better solution too.

**Review by wiidev — APPROVED** [review]

**Comment by Stypox** [comment]
@blackbox87 yeah, do as you said ;-)

**Comment by B0pol** [comment]
thank you

**Comment by avently** [comment]
> thank you

You're welcome. Can wait to see a fix for centering from @blackbox87
